// @tatumio/tatum/dist/src/dto/AddressEventNotificationChain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressEventNotificationChain = void 0;
var AddressEventNotificationChain;
(function (AddressEventNotificationChain) {
    AddressEventNotificationChain["ETH"] = "ETH";
    AddressEventNotificationChain["SOL"] = "SOL";
    AddressEventNotificationChain["XRP"] = "XRP";
    AddressEventNotificationChain["MATIC"] = "MATIC";
    AddressEventNotificationChain["CELO"] = "CELO";
    AddressEventNotificationChain["KLAY"] = "KLAY";
    AddressEventNotificationChain["BTC"] = "BTC";
    AddressEventNotificationChain["LTC"] = "LTC";
    AddressEventNotificationChain["BCH"] = "BCH";
    AddressEventNotificationChain["DOGE"] = "DOGE";
    AddressEventNotificationChain["TRON"] = "TRON";
    AddressEventNotificationChain["BSC"] = "BSC";
    AddressEventNotificationChain["TEZOS"] = "TEZOS";
    AddressEventNotificationChain["EON"] = "EON";
})(AddressEventNotificationChain = exports.AddressEventNotificationChain || (exports.AddressEventNotificationChain = {}));
