// @tatumio/tatum/dist/src/dto/Currency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.networkToCurrency = exports.Currency = void 0;
const Network_1 = require("./Network");
var Currency;
(function (Currency) {
    Currency["ETH"] = "ETH";
    Currency["BTC"] = "BTC";
    Currency["DOGE"] = "DOGE";
    Currency["LTC"] = "LTC";
})(Currency = exports.Currency || (exports.Currency = {}));
function networkToCurrency(network) {
    switch (network) {
        case Network_1.Network.ETHEREUM:
        case Network_1.Network.ETHEREUM_SEPOLIA:
            return Currency.ETH;
        case Network_1.Network.BITCOIN:
        case Network_1.Network.BITCOIN_TESTNET:
            return Currency.BTC;
        case Network_1.Network.DOGECOIN:
        case Network_1.Network.DOGECOIN_TESTNET:
            return Currency.DOGE;
        case Network_1.Network.LITECOIN:
        case Network_1.Network.LITECOIN_TESTNET:
            return Currency.LTC;
        default:
            throw new Error(`Unsupported network ${network}`);
    }
}
exports.networkToCurrency = networkToCurrency;
